coral-common/src/main/java/com/linkedin/coral/common/transformers/OperatorTransformer.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright 2017-2023 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.coral.common.transformers;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+
+/**
+ * Abstract class for generic transformations on SqlCalls
+ */
+public abstract class OperatorTransformer {
+  private SqlValidator sqlValidator;
+  private final List<SqlSelect> topSelectNodes = new ArrayList<>();
+
+  public OperatorTransformer() {
+
+  }
+
+  public OperatorTransformer(SqlValidator sqlValidator) {
+    this.sqlValidator = sqlValidator;
+  }
+
+  protected abstract boolean condition(SqlCall sqlCall);
+
+  protected abstract SqlCall transform(SqlCall sqlCall);
+
+  public SqlCall apply(SqlCall sqlCall) {
+    if (sqlCall instanceof SqlSelect) {
+      this.topSelectNodes.add((SqlSelect) sqlCall);
+    }
+    if (condition(sqlCall)) {
+      return transform(sqlCall);
+    } else {
+      return sqlCall;
+    }
+  }
+
+  /**
+   * To get the RelDatatype of a SqlNode, we iterate through `topSelectNodes` in reverse chronological order,
```

Comment:
I feel `reverse chronological` is not capturing the intended ordering. Maybe it will be clearer after we finish reviewing the example below.